### DISCUSSION TITLE
Stitching Schema Merge

### DISCUSSION THREAD

**Discussion by michaelstaib** [ChilliCream/hotchocolate#4892] [Ideas]
# Supergraph

The **supergraph** in Hot Chocolate represents an internal annotated GraphQL schema that is used by the Hot Chocolate gateway to serve up a distributed schema. The supergraph is the output of the schema merge process.

> In a distributed schema we have the following terms to describe the various schemas:
> - super-graph: The internal annotated schema that describes to the stitching engine how data can be resolved from the various downstream services.
> - subgraph: A schema from a downstream service that represents a partial part of the overall distributed graph.
> - distributed-graph: The schema that is publicly exposed by the Hot Chocolate Gateway and represents the schema that was composed of the subgraphs.

```graphql
type Customer {
  id: ID! 
    @ _hc_bind(to: "customer")
    @ _hc_bind(to: "contracts")
  name: String!
    @_hc_bind(to: "customer" as: "_name")
  complexArg(arg:ComplexInputType): String 
    @ _hc_bind(to: "customer")
  consultant: Consultant
    @ _hc_bind(to: "customer")
  contractsLegacy: [Contract!]
    @_hc_fetch(schema: "contract", select: "foo { contracts(customerId:$fields_id) { ... Contract } }")
  contractsLegacy: [Contract!]
    @_hc_delegate(schema: "contract", path: "contracts(customerId:$fields:id)")
  contractIds: [ID!]
    @_hc_computed
    @_hc_dependsOn(fields: "id contracts")
}

schema 
    @_hc_httpClient(name: "customer", baseAddress: "https://...", features: { allowBatching: true })
    @_hc_httpClient(name: "contracts", baseAddress: "https://...")
    @_hc_webSocketClient(name: "contracts", baseAddress: "wss://...") {
    query: Query
}
```

## Directives

super-graph directives are always prefixed with `_hc_` so that we have a smaller risk of collisions with schema directives of the actual schema. 
super-graph annotations are not meant to be written by hand but are added during the schema merge process. With the new schema stitching version, we will allow the super-graph creation to be done by the gateway (like in the current version) or by tooling that runs outside of the gateway and pushes the super-graph to the gateway.

### Bind

The bind directive is the most important directive and annotates where a certain piece of information can be retrieved from. With only the bind directive we can already create a fully qualified distributed schema.

Bind to a schema

```graphql
@ _hc_bind(to: "customer")
```

Alternatively, a field can also be bound under a different name.

```graphql
@ _hc_bind(to: "customer", as: "bar")
```

The bind as directive can be used also on arguments, enum values  and type names.

### Fetch

The fetch directive allows us to formulate a manual query to fetch data from a downstream service. This directive allows us to fuse the federated approach with the stitching approach.
The `... Fields` fragment spread is where we shall inject the request fields.

```graphql
@_hc_fetch(schema: "contract", select: "foo { contracts(customerId:$fields_id, arg1: $argument_arg1) { ... Fields } }", if: NOT_NULL)
```

### Delegate

The delegate directive is used to allow legacy schemas to work with the new stitching engine by supporting the legacy path to refer to a downstream gateway.

```graphql
@_hc_delegate(schema: "contract", path: "contracts(customerId:$fields:id)")
```

### DependsOn

The depends on directive allows us to express dependencies on other fields in or schema. This enforces the execution engine to first resolve the dependency data before being able to resolve the fie actual field.

```graphql
@_hc_dependsOn(fields: "id contracts { id }")
```

### Computed

The computed directive can be used in combination with custom directives of custom middleware and allows for the extensibility of the stitching engine. If a field is marked as computed the stitching engine will not handle it but expect it to be handled. However, computed can be used in combination with dependsOn. 

```graphql
@_hc_computed
```

**Comment by michaelstaib** [comment]
Here are some links for research of other systems:
https://www.apollographql.com/docs/federation/federation-spec/
https://github.com/atlassian-labs/nadel
https://www.graphql-mesh.com

**Comment by michaelstaib** [comment]
The schema merger should get a list of schema documents and merge them into a single schema document.

```csharp
public interface ISchemaMerger 
{
      ISchemaDocument Merge(IEnumerable<ISchemaDocument> schemas)
}
```

There might be rule sets for merging schema documents.
The schema document consists of schema definition syntax nodes plus some more semantics about the schema.

We could either have multiple ISchemaMerger implementations or a single implementation that is initialized with a ruleset.

**Reply by michaelstaib** [reply]
We should move the whole schema merge into a separate package so that we can use it also from console tooling.

`HotChocolate.Stitching.Types`

**Reply by vhatsura** [reply]
How the information about HTTP and WebSocket clients will be provided for SchemaMerger? Each schema will have such information or it will be passed independently? I'd say it should be apart from ISchemaDocument as it looks wrong when downstream service knows about his address.

(it isn't affected the whole idea of super-graph, but influences to SchemaMerger contract)

**Reply by michaelstaib** [reply]
Good question. The schema document definitely does not deal with connection details. So there could be an envelope type with this information. The question here is what files downstream services expose to us. It could be a schema document with some annotations or we leave these things apart.

I think I will draft a small proposal for that and add it to the discussion so that the picture becomes more complete.

**Comment by matt-psaltis** [comment]
Was looking for ways to use the SDL primitives more for the merging definitions. For consideration:

```
enum Feature {
  BATCHING
}

type HttpClientConnection {}
type WebSocketClientConnection {}
union Connections = HttpClientConnection | WebSocketClientConnection

directive @connections on SCHEMA {
  connections: Connections
}

directive @source on TYPE {
  name: String!
  uri: String!
  features: [Feature]
}

extend type HttpClientConnection
@source(
    name: "CustomerService"
    uri: "https://customerservice"
    features: [BATCHING]
) {
}

extend type WebSocketClientConnection
@source(
    name: "ContractService"
    uri: "wss://contractservice"
) {
}

schema @connections(connections: Connections) {
  query: Query
}

type CustomerExtendedDetails {
  id: ID!
  contracts: []
}

type Customer {
  id: ID!
    @_hc_fetch(schema: "ContractService")
    @_hc_fetch(schema: "CustomerService")
  complexArg(arg: ComplexInputType): String 
  contractsLegacy: [Contract!]
    @_hc_fetch(schema: "ContractService", select: foo, variables: [{id: { contractIds: { id } }}])
  contractIds: [ID!]
    @_hc_computed(from: id)
    @_hc_dependsOn( fields: "id contracts")
}

query/fragment? foo(id: UUID!) {
  contracts(customerId: $id) {
    ... Fields
  }
}
```

**Comment by michaelstaib** [comment]
So,

I have already done the following nodes:

**Finished:**

ArgumentNode
BooleanValueNode
ComplexTypeDefinitionNodeBase
DirectiveDefinitionNode
DirectiveNode
DocumentNode
EnumTypeDefinitionNode
EnumTypeDefinitionNodeBase
EnumTypeExtensionNode
EnumValueDefinitionNode
EnumValueNode
FieldDefinitionNode
FieldNode
FragmentDefinitionNode
FragmentSpreadNode
InlineFragmentNode
InputObjectTypeDefinitionNode
InputObjectTypeExtensionNode
InputValueDefinitionNode
InterfaceTypeDefinitionNode
InterfaceTypeExtensionNode
OptionalModifierNode
IntValueNode
FloatValueNode
NamedTypeNode
ListNullabilityNodeTests
ListTypeNode
ListValueNode
NamedSyntaxNode
NameNode
NullValueNode
ObjectValueNode
ObjectFieldNode
ScalarTypeDefinitionNode
ScalarTypeExtensionNode
OperationTypeDefinitionNode
NonNullTypeNode
ObjectTypeDefinitionNode
ObjectTypeExtensionNode
RequiredModifierNode
SchemaCoordinateNode
SchemaDefinitionNode
SchemaDefinitionNodeBase
SchemaExtensionNode
SelectionSetNode
StringValueNode
UnionTypeDefinitionNode
UnionTypeDefinitionNodeBase
UnionTypeExtensionNode
VariableDefinitionNode
VariableNode

A good example is `EnumValueNode`. When doing nodes that inherit from another look at `FieldDefinitionNode` or for a base type `ComplexTypeDefinitionNodeBase`.

For equality of a list of nodes use `EqualityHelper.Equals` and for the hashcode `EqualityHelper.GetHashCode`.

Take 2 or 3 nodes at once per PR. We will do the tests in the second round, I will create a template for the tests.

Here is the list:

OperationDefinitionNode (Vadim)  


Put you names behind the nodes you will be working on.

**Comment by michaelstaib** [comment]
Regarding Tests.

For each node, we now need tests that verify that the equality is correctly applied.

The first test evaluates that the syntax node`s equality is correct when all nodes are in the same location.

```csharp
[Fact]
public void EqualsBooleanValueNode_SameLocation()
{
    // arrange
    var a = new BooleanValueNode(new Location(1, 1, 1, 1), false);
    var b = new BooleanValueNode(new Location(1, 1, 1, 1), false);
    var c = new BooleanValueNode(new Location(1, 1, 1, 1), true);

    // act
    var abResult = a.Equals(b);
    var aaResult = a.Equals(a);
    var acResult = a.Equals(c);
    var aNullResult = a.Equals(default);

    // assert
    Assert.True(abResult);
    Assert.True(aaResult);
    Assert.False(acResult);
    Assert.False(aNullResult);
}
```

The second test ensures that the location has no impact on equality. Two boolean nodes in this case are the same if the parsed boolean value is the same.

```csharp
[Fact]
public void EqualsBooleanValueNode_DifferentLocation()
{
    // arrange
    var a = new BooleanValueNode(new Location(1, 1, 1, 1), false);
    var b = new BooleanValueNode(new Location(2, 2, 2, 2), false);
    var c = new BooleanValueNode(new Location(3, 3, 3, 3), true);

    // act
    var abResult = a.Equals(b);
    var aaResult = a.Equals(a);
    var acResult = a.Equals(c);
    var aNullResult = a.Equals(default);

    // assert
    Assert.True(abResult);
    Assert.True(aaResult);
    Assert.False(acResult);
    Assert.False(aNullResult);
}
```

Lastly, we want to ensure that the location has no impact on the hashCode.

```csharp
[Fact]
public void CompareGetHashCode_With_Location()
{
    // arrange
    var a = new BooleanValueNode(new(1, 1, 1, 1), false);
    var b = new BooleanValueNode(new(2, 2, 2, 2), false);
    var c = new BooleanValueNode(new(1, 1, 1, 1), true);
    var d = new BooleanValueNode(new(2, 2, 2, 2), true);

    // act
    var aHash = a.GetHashCode();
    var bHash = b.GetHashCode();
    var cHash = c.GetHashCode();
    var dHash = d.GetHashCode();

    // assert
    Assert.Equal(aHash, bHash);
    Assert.NotEqual(aHash, cHash);
    Assert.Equal(cHash, dHash);
    Assert.NotEqual(aHash, dHash);
}
```

I will add another example for a more complex node.

**Reply by michaelstaib** [reply]
Here the example with the `ArgumentNode`.

```csharp
[Fact]
    public void Equals_With_Same_Location()
    {
        // arrange
        var a = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(123));
        var b = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(123));
        var c = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(567));

        // act
        var abResult = a.Equals(b);
        var aaResult = a.Equals(a);
        var acResult = a.Equals(c);
        var aNullResult = a.Equals(default);

        // assert
        Assert.True(abResult);
        Assert.True(aaResult);
        Assert.False(acResult);
        Assert.False(aNullResult);
    }

    [Fact]
    public void Equals_With_Different_Location()
    {
        // arrange
        var a = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(123));
        var b = new ArgumentNode(new Location(2, 2, 2, 2), new("aa"), new IntValueNode(123));
        var c = new ArgumentNode(new Location(3, 3, 3, 3), new("aa"), new IntValueNode(567));

        // act
        var abResult = a.Equals(b);
        var aaResult = a.Equals(a);
        var acResult = a.Equals(c);
        var aNullResult = a.Equals(default);

        // assert
        Assert.True(abResult);
        Assert.True(aaResult);
        Assert.False(acResult);
        Assert.False(aNullResult);
    }

    [Fact]
    public void GetHashCode_With_Location()
    {
        // arrange
        var a = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(123));
        var b = new ArgumentNode(new Location(2, 2, 2, 2), new("aa"), new IntValueNode(123));
        var c = new ArgumentNode(new Location(1, 1, 1, 1), new("aa"), new IntValueNode(567));
        var d = new ArgumentNode(new Location(2, 2, 2, 2), new("aa"), new IntValueNode(567));

        // act
        var aHash = a.GetHashCode();
        var bHash = b.GetHashCode();
        var cHash = c.GetHashCode();
        var dHash = d.GetHashCode();

        // assert
        Assert.Equal(aHash, bHash);
        Assert.NotEqual(aHash, cHash);
        Assert.Equal(cHash, dHash);
        Assert.NotEqual(aHash, dHash);
    }
```

**Reply by michaelstaib** [reply]
No need to say that we only need tests for the actual nodes not the base types.

**Reply by tsinghammer** [reply]
I've added a new test project to the solution HotChocolate.Language.SyntaxTree.Tests.csproj
Is there already one?

**Reply by tsinghammer** [reply]
Thanks for the example!

**Reply by michaelstaib** [reply]
At the moment all is in the `HotChocolate.Language.Tests.csproj`. But its actually cleaner if we have a separate project ... so once we have merged your I will move over my tests into your new project.

**Reply by michaelstaib** [reply]
I created a helper class for the tests with the locations:

```csharp
namespace HotChocolate.Language;

internal static class TestLocations
{
    public static Location Location1 { get; } = new Location(1, 1, 1, 1);
    public static Location Location2 { get; } = new Location(2, 2, 2, 2);
    public static Location Location3 { get; } = new Location(3, 3, 3, 3);
}
```
